File: baekjoon/src/dynamic1/No2579.java
package dynamic1;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/2579
 */
public class No2579 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static StringTokenizer tokenizer;

    public static int numberOfInput;
    public static int[] scores;
    public static int[] dp;

    public static void setValues() throws IOException {
        numberOfInput = Integer.parseInt(reader.readLine());
        dp = new int[numberOfInput + 1];
        scores = new int[numberOfInput + 1];
        for (int i = 1; i < numberOfInput + 1; i++)
            scores[i] = Integer.parseInt(reader.readLine());
    }

    public static int getMaxScore() {
        int length = scores.length;
        dp[1] = scores[1];
        if (numberOfInput >= 2)
            dp[2] = scores[1] + scores[2];
        for (int i = 3; i < length; i++)
            dp[i] = Math.max(dp[i - 2], dp[i - 3] + scores[i - 1]) + scores[i];
        return dp[length - 1];
    }

    public static void main(String[] args) throws IOException {
        setValues();
        writer.write(String.valueOf(getMaxScore()));
        writer.flush();
    }
}

File: yourssu/src/week8/No2579.kt
package week8

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/2579
 */
private val reader = BufferedReader(InputStreamReader(System.`in`))
private val writer = BufferedWriter(OutputStreamWriter(System.out))

fun main() {
    val numberOfInput = reader.readLine().toInt()
    val stairs = IntArray(301)
    for (i in 1 .. numberOfInput) {
        stairs[i] = reader.readLine().toInt()
    }

    val dp = IntArray(301)
    dp[1] = stairs[1]
    dp[2] = stairs[1] + stairs[2]
    for (i in 3 until stairs.size) {
        dp[i] = dp[i - 2].coerceAtLeast(dp[i - 3] + stairs[i - 1]) + stairs[i]
    }

    writer.write("${dp[numberOfInput]}")
    writer.flush()
}